chain/beacon/sync_manager.go

```diff
@@ -0,0 +1,294 @@
+// round. Depending on the current state of the syncing process, there might not
+// be a new process starting (for example if we already have the round
+// requested). upTo == 0 means the syncing process goes on forever.
+func (s *SyncManager) RequestSync(nodes []net.Peer, upTo uint64) {
+	s.newReq <- requestInfo{
+		nodes: nodes,
+		upTo:  upTo,
+	}
+}
+
+func (s *SyncManager) Run() {
+	// tracks the time of the last round we successfully synced
+	lastRoundTime := 0
+	// the context being used by the current sync process
+	lastCtx, cancel := context.WithCancel(context.Background()) // nolint
+	for {
+		select {
+		case request := <-s.newReq:
+			// check if the request is still valid
+			last, err := s.store.Last()
+			if err != nil {
+				s.log.Debug("sync_manager", "unable to fetch from store", "err", err)
+				continue
+			}
+			// do we really need a sync request ?
+			if request.upTo > 0 && last.Round >= request.upTo {
+				s.log.Debug("sync_manager", "request already filled", "last", last.Round, "request", request.upTo)
+				continue
+			}
+			// check if it's been a while we haven't received a new round from
+			// sync. Either there is a sync in progress but it's stuck, so we
+			// quit it and start a new one, or there isn't and we start one.
+			// We always give a delay of a few periods since the one next to "now"
+			// might not be exactly ready yet so only after a few periods we know we
+			// must have gotten some data.
+			upperBound := lastRoundTime + int(s.period.Seconds())*s.factor
+			if upperBound < int(s.clock.Now().Unix()) {
+				// we haven't received a new block in a while
+				// -> time to start a new sync
+				cancel()
+				lastCtx, cancel = context.WithCancel(context.Background())
+				go s.sync(lastCtx, request)
+			}
+		case <-s.newSync:
+			// just received a new beacon from sync, we keep track of this time
+			lastRoundTime = int(s.clock.Now().Unix())
+		case <-s.done:
+			s.log.Info("sync_manager", "exits")
+			cancel()
+			return
+		}
+	}
+}
+
+func (s *SyncManager) sync(ctx context.Context, request requestInfo) {
+	s.log.Debug("sync_manager", "starting new sync", "up_to", request.upTo, "nodes", peersToString(request.nodes))
+	// shuffle through the nodes
+	for _, n := range rand.Perm(len(request.nodes)) {
+		node := request.nodes[n]
+		if s.tryNode(ctx, request.upTo, node) {
+			return
+		}
+	}
+	s.log.Debugw("Tried all nodes without success", "sync_manager", "failed sync")
+}
+
+// tryNode tries to sync up with the given peer up to the given round, starting
+// from the last beacon in the store. It returns true if the objective was
+// reached (store.Last() returns upTo) and false otherwise.
+func (s *SyncManager) tryNode(global context.Context, upTo uint64, n net.Peer) bool {
+	// we put a cancel to still keep the global context open but stop with this
+	// peer if things go sideway
+	cnode, cancel := context.WithCancel(global)
```

Comment:
Should we have a context with a timeout here, or do we rely on the GRPC client to  timeout?

Reply:
I think we started trying to plumb life-cycles through a bit more previously, but didn't really get them complete so I don't think we need to require that as part of this PR either.